### PR TITLE
Improvement/custom project creation

### PR DESCRIPTION
**Pull request by adrianherrera**
refactoring of project creation (again 🙂). I want/need the ability to create "custom projects" that are outside the standard S2E Windows/Linux/CGC project, and the current project + config class made this a bit clunky.

I think the new structure simplifies this (mainly because it removes the config classes) and it has the bonus of less mutable state (which I like as an occasional functional programmer 🙂).

Let me know what you think

### CODE REVIEW

**Comment by vitalych** [comment]
LGTM! Some high-level remarks:
- pylint reports errors (see Jenkins)
- Did you try all creating all types of projects / device drivers, etc?
- I think at this point it would be a really good idea to add some basic unit/integration tests (i.e., create a bunch of projects and check that the produced files look ok) and get some code coverage support. Could use pytest/mock for that. I am a bit afraid to merge this without testing.

**Comment by adrianherrera** [comment]
Yup, what you said makes sense. I've done manual testing of project creation, but I'll look at doing something more automated over the weekend. It's a (relatively) large change!

**Comment by adrianherrera** [comment]
Okay, I've done a lot more refactoring. The main change is in the `AbstractProject` class: project creation is split into two abstract methods: `_make_project_config` and `_create_project`.

`_make_project_config` creates a dict that describes everything about the project, and `_create_project` actually writes everything to disk. This is a nicer separation of concerns and makes it easier to test: now I just have to write tests for `_make_project_config` that checks that everything is configured correctly.

**Comment by adrianherrera** [comment]
Basic project creation tests added. This is ready to be reviewed. It would be good to add tests for coverage, parsers, etc. But as these do not pertain to this PR (and I only have 2 days before I go on holidays for some weeks), I'll save it for another time...

**Comment by vitalych** [comment]
Could you update the readme to explain how to run the tests?

File: s2e_env/commands/new_project.py [hunk 1]
```diff
@@ -300,7 +301,18 @@ def add_arguments(self, parser):
                                  'exists, replace it')
 
     def handle(self, *args, **options):
-        if options['target']:
-            _handle_with_file(*args, **options)
+        # The 'project' option is not exposed as a command-line argument: it is
+        # typically used when creating a custom project programatically. It
+        # provides a class that is instantiated with the current command-line
+        # arguments and options
+        proj_class = options.get('project')
```
Comment (vitalych) [inline s2e_env/commands/new_project.py]: It's not clear to me how I can use this feature. A note in the readme would help.

**Comment by adrianherrera** [comment]
> Could you update the readme to explain how to run the tests?

Done

File: s2e_env/commands/new_project.py (hunk 1 above)
Comment (adrianherrera) [reply s2e_env/commands/new_project.py]: Done

**Comment by vitalych** [comment]
Looks like there are some unicode errors
```
AssertionError: Lists differ: ['/home/vitaly/s2e/s2e/s2e-env... != ['/home/vitaly/s2e/s2e/s2e-env...

First differing element 1:
u'/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanuser.exe'
'/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanner.sys'

  ['/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanner.inf',
+  '/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanner.sys',
-  u'/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanuser.exe',
?  -                                                                                      ^

+  '/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanuser.exe']
?                                                                                        ^

-  u'/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/../../data/scanner.sys']

----------------------------------------------------------------------
```

**Comment by vitalych** [comment]
I had to install mock manually, for some reason it didn't get picked up by the normal installation process with `pip install --process-dependency-links -e .` Maybe there is another command for that.

```
======================================================================
ERROR: tests.commands.project_creation.test_windows_project (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------                                                                                                              
ImportError: Failed to import test module: tests.commands.project_creation.test_windows_project
Traceback (most recent call last):                                   
  File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests            
    module = self._get_module_from_name(name)
  File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name             
    __import__(name)                                                                                                                     
  File "/home/vitaly/s2e/s2e/s2e-env/tests/commands/project_creation/test_windows_project.py", line 28, in <module>
    from mock import MagicMock                                   
ImportError: No module named mock 
```

**Comment by adrianherrera** [comment]
> I had to install mock manually, for some reason it didn't get picked up by the normal installation process with pip install --process-dependency-links -e .

Maybe pip treats `test_requires` differently in `setup.py`. I have no idea...

> Looks like there are some unicode errors

That's interesting! I don't get any unicode errors, everything passes for me.

**Comment by adrianherrera** [comment]
Oh, that's weird. Does the inf driver analyze and extract files in a non-deterministic order?

I think the correct solution would be to use `self.assertCountEqual`. Despite it's name, it compares the contents of a list, ignoring order (source: https://docs.python.org/3.2/library/unittest.html#unittest.TestCase.assertCountEqual)

**Comment by adrianherrera** [comment]
Wait, I don't think `assertCountEqual` exists in Python 2 :(

**Comment by adrianherrera** [comment]
Pushed a new version. Used `assertItemsEqual` for that test (https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.assertItemsEqual). Breakfast time 🥣

**Comment by vitalych** [comment]
Now it works. Regarding non-deterministic order, I think this is the issue:
https://github.com/S2E/s2e-env/blob/improvement/custom-project-creation/s2e_env/commands/new_project.py#L141
I have different paths that get sorted differently.

**Comment by adrianherrera** [comment]
Ah yes of course, the `set` will throw away the ordering.

**Pull request merged**